quattor/functions/validation.pan

```diff
@@ -139,3 +139,15 @@ function valid_absolute_file_paths = {
     };
     true;
 };
+
+@documentation{
+    desc = Function to validate whether the parameter received is a valid username or not; a valid username is considered to only be comprised of alpha numeric characters, plus a few special ones which are: .+:-
+}
+function valid_user_name = {
+    user = to_string(ARGV[0]);
+    if(!match(user, "^[A-Za-z0-9._:-]+$")) {
```

Comment:
and the usernames can start with the funky stuff too?

Reply:
True, it shouldn't. I actually converted a regex that Morgan Stanley was using internally to a function. Not sure why characters such as `.` or `:` are authorized. Will check. Might change that to the more standard regex for usernames: `^[A-Za-z_]([A-Za-z0-9_-]{0,31}|[A-Za-z0-9_-]{0,30}\$)$`

Reply:
usernames can start with digits (i think posix allows usernames to be digits only, but lets not go there ;)

Reply:
If we want to be entirely accurate, we might as well support actually all usernames that can be accepted, from http://manpages.ubuntu.com/manpages/bionic/man8/useradd.8.html (and other resources):

```
       It is usually recommended to only use usernames that begin with a lower case letter or an
       underscore, followed by lower case letters, digits, underscores, or dashes. They can end
       with a dollar sign. In regular expression terms: [a-z_][a-z0-9_-]*[$]?

       On Debian, the only constraints are that usernames must neither start with a dash ('-')
       nor plus ('+') nor tilde ('~') nor contain a colon (':'), a comma (','), or a whitespace
       (space: ' ', end of line: '\n', tabulation: '\t', etc.). Note that using a slash ('/') may
       break the default algorithm for the definition of the user's home directory.

       Usernames may only be up to 32 characters long.
```

The lack of limitation is linked to the fact that PAM allows pretty much anything to be used, while the old passwd was more limitating. Maybe @ned21 would have a better idea as to why this regex is currently used at Morgan Stanley, and what would be acceptable not to break anything for us nor for others ?

Reply:
hmm.. I probably don't know the real history, git blame on the internal repository might yield some light but I doubt it. 

I would go with what is well documented for e.g. ubuntu (can we check RH too?) The one caveat is that I know `.` is definitely required because in some places we ended up using `firstname.lastname` as a userid. I am guessing that also exceeded the 32 character limit that Debian/Ubuntu imposes. 

I would randomly guess that `:` might have been considered valid because in various configuration components, passing `user:group` was a way to set the user and group of a file. In fact, @XaF you will probably find that the there is template code internally that says a field is a "user" but actually accepts `user:group` syntax to allow for the automatic provisioning of directories based on which users may login to a host. (Sorry I forgot about this when discussing the code internally, it was only when comparing the two regexes that I spotted the potential significance of them.)

Reply:
From what I understand of the `useradd` man page, the only real "official" limitation is for the first char (Debian/Ubuntu) and for the username length of 32 characters. The rest is only a recommendation. Moreover, as I was saying before, the generalisation of using PAM for authentication made those limitations even blurrier, and we now only have recommendations that apply. Pretty much every system can have a different setup and different rules for usernames (from starting and ending characters to username length).

It seems to be a bigger issue than one that could be tackled in a general way. I feel that if we are very specific about that internally, we might want to keep that type for ourselves and not burden the community with something that will only be useful to us. Thoughts?

Reply:
If this validation result in allowing usernames that other sites would consider invalid then I agree: let's defer it to each site to define their own. It would be nice to document that decision somewhere to prevent a re-run of this conversation in the future but I'm not sure where would make sense. This would be a clear use case for being able to override or add to types in PAN: we could then define `type valid_username` with a very basic string type/minimal regex, and comment that this is designed to be overridden by site-specific policies.

Reply:
I feel that's a good idea, this would allow to standardize the existence of that variable if we need it in other generic templates, while still keeping it generic enough from the beginning, and site-specific as we recommend to override it.

What would you feel would be a good default value for that? I think that going for the recommended regex defined in the `useradd` man page (that I quoted earlier) might be interesting here. Or would you prefer a regex that allows everything? `^. *$`. Thoughts @ned21 @stdweird @jrha ?